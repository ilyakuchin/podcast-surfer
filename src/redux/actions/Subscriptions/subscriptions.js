import axios from 'axios';
import { buildPodcastRequestUrl } from '../../../helpers/urls';

export const FETCH_SUBSCRIPTIONS_REQUEST = 'FETCH_SUBSCRIPTIONS_REQUEST';
export const FETCH_SUBSCRIPTIONS_SUCCESS = 'FETCH_EPISODE_SUCCESS';
export const FETCH_SUBSCRIPTIONS_FAILURE = 'FETCH_SUBSCRIPTIONS_FAILURE';

export function fetchSubscriptionsRequest() {
  return { type: FETCH_SUBSCRIPTIONS_REQUEST, isFetching: true };
}

export function fetchSubscriptionsSuccess(subscriptions) {
  return {
    type: FETCH_SUBSCRIPTIONS_SUCCESS,
    isFetching: false,
    subscriptions
  };
}

export function fetchSubscriptionsFailure(subscriptions) {
  return { type: FETCH_SUBSCRIPTIONS_FAILURE, subscriptions };
}

export function fetchSubscriptions(subscriptions) {
  return dispatch => {
    dispatch(fetchSubscriptionsRequest());

    const promises = [];

    for (let i = 0; i < subscriptions.length; i += 1) {
      promises.push(axios.get(buildPodcastRequestUrl(subscriptions[i])));
    }
    return Promise.all(promises).then(res => {
      dispatch(fetchSubscriptionsSuccess(res.map(item => item.data)));
    });
  };
}
