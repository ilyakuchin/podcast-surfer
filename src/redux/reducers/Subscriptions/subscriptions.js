import {
  FETCH_SUBSCRIPTIONS_REQUEST,
  FETCH_SUBSCRIPTIONS_SUCCESS,
  FETCH_SUBSCRIPTIONS_FAILURE
} from '../../actions/Subscriptions/subscriptions';

export default function subscriptions(
  state = { isFetching: false, subscriptions: [] },
  action
) {
  switch (action.type) {
    case FETCH_SUBSCRIPTIONS_REQUEST:
      return {
        ...state,
        isFetching: action.isFetching
      };
    case FETCH_SUBSCRIPTIONS_SUCCESS:
      return {
        ...state,
        isFetching: action.isFetching,
        subscriptions: action.subscriptions
      };
    case FETCH_SUBSCRIPTIONS_FAILURE:
      return {
        ...state,
        isFetching: action.isFetching,
        errorMessage: action.errorMessage
      };
    default:
      return state;
  }
}
