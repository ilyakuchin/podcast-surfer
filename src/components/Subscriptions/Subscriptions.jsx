import React, { useEffect } from 'react';
import { connect } from 'react-redux';
import PodcastList from '../PodcastList/PodcastList';
import { fetchCurrentPodcast } from '../../redux/actions/CurrentPodcast/currentPodcast';
import { fetchSubscriptions } from '../../redux/actions/Subscriptions/subscriptions';

export function Subscriptions({
  isFetching,
  subscriptions,
  jwt,
  fetchCurrentPodcastConnect,
  fetchSubscriptionsConnect
}) {
  useEffect(() => {
    fetchSubscriptionsConnect();
  }, [fetchSubscriptionsConnect]);

  return (
    <PodcastList
      jwt={jwt}
      podcasts={subscriptions}
      isFetching={isFetching}
      fetchCurrentPodcast={fetchCurrentPodcastConnect}
    />
  );
}

const mapStateToProps = state => {
  return {
    subscriptions: state.subscriptions.podcasts,
    isFetching: state.subscriptions.isFetching,
    jwt: state.userInfo.jwt
  };
};

const dispatchToProps = dispatch => {
  return {
    fetchCurrentPodcastConnect: (rss, jwt, id) =>
      dispatch(fetchCurrentPodcast(rss, jwt, id)),
    fetchSubscriptionsConnect: () => dispatch(fetchSubscriptions())
  };
};

const ConnectedSubscriptions = connect(
  mapStateToProps,
  dispatchToProps
)(Subscriptions);

export default ConnectedSubscriptions;
