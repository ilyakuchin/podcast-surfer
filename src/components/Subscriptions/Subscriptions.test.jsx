import React from 'react';
import Adapter from 'enzyme-adapter-react-16';
import { configure, shallow } from 'enzyme';
import toJson from 'enzyme-to-json';
import { Subscriptions } from './Subscriptions';

configure({ adapter: new Adapter() });

describe('Test Subscriptions component', () => {
  test('should render correctly', () => {
    expect(
      toJson(
        shallow(
          <Subscriptions
            subscriptions={[]}
            subscriptionUrls={[]}
            isFetching={false}
            jwt='jwt'
            fetchSubscriptionsConnect={() => {}}
          />
        )
      )
    ).toMatchSnapshot();
  });
});
