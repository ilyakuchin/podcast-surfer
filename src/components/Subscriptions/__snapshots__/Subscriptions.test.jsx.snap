// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Subscriptions component should render correctly 1`] = `
<div>
  <Header
    as="h2"
    textAlign="center"
  >
    There is no subscriptions yet
  </Header>
  <styled.div>
    <Connect(PodcastList)
      isFetching={false}
      podcasts={Array []}
    />
  </styled.div>
</div>
`;
